man/utop.1
\" utop.1
\" ------
\" Copyright : (c) 2011, Jeremie Dimino <jeremie@dimino.org>
\" Licence   : BSD3
\"
\" This file is a part of utop.

.TH UTOP 1 "July 12, 2026"

.SH NAME
utop \- Universal toplevel for OCaml

.SH SYNOPSIS
.B utop
[
.I options
]
[
.I object-files
]
[
.I script-file
]

.SH DESCRIPTION

.B utop
is an enhanced toplevel for OCaml with many features, including context
sensitive completion.

When you start
.B utop
what you see is the prompt followed by a bar containing words. This is
the completion bar: it contains the possible completion and is updated
as you type. The highlighted word in the completion bar is the
selected word. You can navigate using the keys \fBM-Left\fR and \fBM-Right\fR
and select one completion using \fBM-down\fR. Here \fBM\fR represents the meta key,
which is \fBAlt\fR on most systems. You can configure these bindings in the file
.I ~/.config/lambda-term-inputrc
- see
.BR lambda-term-inputrc (5)
for details.

\fButop\fR supports completion on:

        * directives and directive arguments
        * identifiers
        * record fields
        * variants
        * function labels
        * object methods

Colors are by default configured for terminals with dark colors, such
as white on black, so the prompt may look too bright on light-colored
terminals. You can change that by setting the color profile of
utop. To do that, type:

        UTop.set_profile UTop.Light;;

You can then add this line to your
.I ~/.config/utop/init.ml
file.

To turn off \fButop\fR's advanced prompt features, add the following to \fIinit.ml\fR
to turn off respectively (a) colors and the upper information line, and
(b) the lower boxed list of possible completions:

        #utop_prompt_dummy;;
        UTop.set_show_box false

To disable \fButop\fR's welcome message, add the folowing to \fIinit.ml\fR:

        UTop.set_show_welcome false;;

You can enable basic syntax highlighting in utop by writing a
.I $XDG_CONFIG_HOME/utop/utoprc
file (usually
.IR ~/.config/utop/utoprc ).
See
.BR utoprc (5)
for that.

Vi edit mode is enabled by the command

        #edit_mode_vi

It currently supports three vi modes: normal, insert, visual mode, and you can
get/set content with vim-like registers.

.I utop.el
is a package that provides
.B utop
integration with Emacs. The package allows you to run
.B utop
inside Emacs and to evaluate code in it straight from your source buffers
(with the help of \fIutop-minor-mode\fR).
The recommended way to install
.I utop.el
is via Emacs's built-in package manager \fBpackage.el\fR.
More detailed installation and configuration instructions may be found
on the project's code repository at \fIhttps://github.com/ocaml-community/utop\fR.

You can start \fButop\fR inside Emacs with \fBM-x utop\fR.

The default install also has a minor mode with the following key bindings:

        \fBC-c C-s\fR Start a utop buffer (\fIutop\fR)
        \fBC-x C-e\fR Evaluate the current phrase (\fIutop-eval-phrase\fR)
        \fBC-x C-r\fR Evaluate the selected region (\fIutop-eval-region\fR)
        \fBC-c C-b\fR Evaluate the current buffer (\fIutop-eval-buffer\fR)
        \fBC-c C-k\fR Kill a running utop process (\fIutop-kill\fR)
        \fBC-c C-z\fR Switch to utop process (\fIutop-switch-to-repl\fR)

then you can run \fButop\fR by pressing \fBM-x\fR and typing "utop". \fButop\fR supports
completion in Emacs mode. Just press \fBTab\fR to complete a word. You can
also integrate it with the tuareg, caml, reason or typerex mode. Other
modes can register support via \fIutop-mode-compat-alist\fR. For that add
the following lines to your
.I ~/.emacs
file:

        (autoload 'utop-minor-mode "utop" "Minor mode for utop" t)
        (add-hook 'tuareg-mode-hook 'utop-minor-mode)

.SH OPTIONS
See
.B utop --help
for the full list of available options. There is considerable overlap
with options available for
.BR ocaml (1).

A commonly used option is
\fB-require\fI package\fR
to load \fIpackage\fR into the execution environment. It is equivalent
to using \fB#require\fR from inside
.BR utop (1).

.PP
.TP
.BI -absname
Show absolute filenames in error message.
.TP
.BI -I " dir"
Add \fIdir\fR to the list of include directories.
.TP
.BI -init " file"
Load \fIfile\fR instead of the default init file.
.TP
.BI -labels
Use commuting label mode.
.TP
.BI -no-app-funct
Deactivate applicative functors.
.TP
.BI -noassert
Do not compile assertion checks.
.TP
.BI -nolabels
Ignore non-optional labels in types.
.TP
.BI -nostdlib
Do not add the default directory to the list of include directories.
.TP
.BI -ppx " command"
Pipe abstract syntax trees through the preprocessor \fIcommand\fR.
.TP
.BI -principal
Check principality of type inference.
.TP
.BI -safe-string
Make strings immutable (only available with OCaml < 5.0).
.TP
.BI -short-paths
Shorten paths in types (the default).
.TP
.BI -no-short-paths
Do not shorten paths in types.
.TP
.BI -rectypes
Allow arbitrary recursive types.
.TP
.BI -stdin
Read script from standard input.
.TP
.BI -strict-sequence
Left-hand part of a sequence must have type unit.
.TP
.BI -unsafe
Do not compile bounds checking on array and string access.
.TP
.BI -version
Print version and exit.
.TP
.BI -vnum
Print version number and exit.
.TP
.BI -w " list"
Enable or disable warnings according to \fIlist\fR.
.TP
.BI -warn-error " list"
Enable or disable error status for warnings according to \fIlist\fR.
See option \fB-w\fR for the syntax of \fIlist\fR.
Default setting is \fB-a+31\fR.
.TP
.BI -warn-help
Show description of warning numbers.
.TP
.BI -emacs
Run in emacs mode.
.TP
.BI -hide-reserved
Hide identifiers starting with a '_' (the default).
.TP
.BI -show-reserved
Show identifiers starting with a '_'.
.TP
.BI -no-implicit-bindings
Don't add implicit bindings for expressions (the default).
.TP
.BI -implicit-bindings
Add implicit bindings: \fIexpr\fR;; -> let _0 = \fIexpr\fR;;
.TP
.BI -no-autoload
Disable autoloading of files in 
.I $OCAML_TOPLEVEL_PATH/autoload.
.TP
.BI -require " package"
Load this package.
.TP
.BI -dparsetree
Dump OCaml AST after rewriting.
.TP
.BI -dsource
Dump OCaml source after rewriting.
.TP
.BI -help
Display this list of options.
.TP
.BI --help
Display this list of options.

.SH FILES
.I $XDG_CONFIG_HOME/utop/init.ml
.RS
The initialization file of the toplevel (usually
.IR ~/.config/utop/init.ml ).
A
.I .ocamlinit
file in the current directory or a
.I ~/.ocamlinit
file is also picked up.
.RE
.I $XDG_CONFIG_HOME/utop/utoprc
.RS
The configuration file for utop (usually
.IR ~/.config/utop/utoprc ).
See
.BR utoprc (5).
The legacy location
.I ~/.utoprc
is still honored if that file exists.
.RE
.I $XDG_STATE_HOME/utop/utop-history
.RS
The history file (usually
.IR ~/.local/state/utop/utop-history ).
The legacy location
.I ~/.utop-history
is still honored if that file exists.
.RE
.I ~/.config/lambda-term-inputrc
.RS
The file containing key bindings. See
.BR lambda-term-inputrc (5).

.SH AUTHOR
Jérémie Dimino <jeremie@dimino.org>

.SH "SEE ALSO"
.BR utoprc (5),
.BR lambda-term-inputrc (5),
.BR ocaml (1).


CHANGES
===== 1.2.1 (2012-07-31) =====

  * fix: do not expunge Toploop
  * install a non-expunged version of utop: utop-full

===== 1.2 (2012-07-30) =====

  * ocaml 4.00 compatibility
  * prevent findlib from being initialized twice
  * better highlighting of errors
  * automatically insert Lwt_main.run for
    toplevel expressions of type 'a Lwt.t
  * better camlp4 support
    ** parse quotations and antiquotations to
       handle completion inside them
    ** better support revised syntax
  * emacs mode improvements
    ** various fixes
    ** highlight errors
    ** add a menu
    ** add interactive list of findlib packages
    ** packages can be pre-loaded via the file variable
       "utop-package-list"
    ** better tuareg integration
    ** typerex integration
    ** allow to complete using the toplevel environment
       in a tuareg buffer
    ** allow to change the utop command
    ** use the same history as the terminal mode
    ** follow output of ocaml

===== 1.1 (2010-08-06) =====

  * add completion on labels
  * add completion on methods
  * smarter completion on record fields
  * fix a bug in the lexer
  * improvement for the emacs mode:
    ** now pressing Tab really complete input
    ** when sending input from a tuareg buffer, the cursor follow the
       end of buffer in all utop windows
    ** fix usage of threads
  * add help
  * add manual pages
  * show more information in the prompt:
    ** show the current value of the macro counter
    ** show the nnumber of key pressed since the beginning of a macro
       when recording a macro
    ** show intermediate key sequence
  * better support for light colors terminals
  * add colors for module name and directives
  * add UTop.smart_accept to send only lines terminating with a ";;" token
  * search for compiler libraries at configure time
  * add a script to install compiler libraries
  * fix compatibility with ocaml 3.13


src/utop.el
;; utop.el
;; -------
;; Copyright : (c) 2011, Jeremie Dimino <jeremie@dimino.org>
;; Licence   : BSD3
;;
;; This file is a part of utop.

;; +-----------------------------------------------------------------+
;; | Customizable variables                                          |
;; +-----------------------------------------------------------------+

(defgroup utop nil
  "A toplevel for the ocaml programming language which interact
with Emacs to provide an enhanced environment."
  :tag "The Caml Emacs-Lisp Toplevel"
  :version "1.0"
  :group 'applications)

(defcustom utop-command "utop-emacs"
  "The command to execute for utop."
  :type 'string
  :group 'utop)

(defcustom utop-prompt 'utop-default-prompt
  "The function which create the prompt for utop."
  :type 'function
  :group 'utop)

(defcustom utop-mode-hook nil
  "A hook that gets run when `utop-mode' is entered."
  :type 'hook
  :group 'utop)

(defcustom utop-exit-hook nil
  "A hook that is run whenever `utop' is exited.
This hook is only run if exiting actually kills the buffer."
  :type 'hook
  :group 'utop)

(defface utop-prompt
  '((t (:foreground "Cyan1")))
  "The face used to highlight the prompt."
  :group 'utop)

(defface utop-stdout
  nil
  "The face used to highlight messages comming from stdout."
  :group 'utop)

(defface utop-stderr
  nil
  "The face used to highlight messages commong from stderr."
  :group 'utop)

(defface utop-frozen
  '((t (:bold t)))
  "The face used to highlight text that has been sent to utop.")

;; +-----------------------------------------------------------------+
;; | Constants                                                       |
;; +-----------------------------------------------------------------+

(defconst utop-buffer-name "*utop*"
  "The name of the buffer utop is running on.")

;; +-----------------------------------------------------------------+
;; | Variables                                                       |
;; +-----------------------------------------------------------------+

(defvar utop-process nil
  "The Lisp-object for the utop sub-process")

(defvar utop-mode-map nil
  "The utop local keymap.")

(defvar utop-prompt-min 0
  "The point at the beginning of the current prompt.")

(defvar utop-prompt-max 0
  "The point at the end of the current prompt.")

(defvar utop-last-prompt 0
  "The contents of the last displayed prompt.")

(defvar utop-output ""
  "The output of the utop sub-process not yet processed.")

(defvar utop-command-number 0
  "The number of the current command.")

(defvar utop-history nil
  "The history of typed command.")

(defvar utop-history-prev nil
  "The history before the cursor.")

(defvar utop-history-next nil
  "The history after the cursor.")

(defvar utop-pending nil
  "The text not yet added to the history.")

(defvar utop-completion nil
  "Current completion.")

(defvar utop-move-to-end-of-buffer nil
  "Whether to move the point to the end of prompt after
displaying the prompt")

;; +-----------------------------------------------------------------+
;; | Utils                                                           |
;; +-----------------------------------------------------------------+

(defun utop-add-text-properties-rear-nonsticky (start end properties nonsticky-properties &optional object)
  "Same as ``add-text-properties'' but put the last character in
non-sticky mode."
  (when (< start end)
    ;; Put everything between start and end-1 in sticky read-only mode
    (add-text-properties start (- end 1) properties object)
    ;; Put the last character in non-sticky mode
    (add-text-properties (- end 1) end
                         (append
                          properties
                          (list 'rear-nonsticky nonsticky-properties))
                         object)))

;; +-----------------------------------------------------------------+
;; | Prompt                                                          |
;; +-----------------------------------------------------------------+

(defun utop-default-prompt ()
  "The default prompt function."
  (let ((prompt (format "utop[%d]> " utop-command-number)))
    (add-text-properties 0 (length prompt) '(face utop-prompt) prompt)
    prompt))

;; +-----------------------------------------------------------------+
;; | History                                                         |
;; +-----------------------------------------------------------------+

(defun utop-history-goto-prev ()
  "Go to the previous entry of the history."
  (interactive)
  (unless (null utop-history-prev)
    (with-current-buffer utop-buffer-name
      ;; Push current input after the history cursor
      (push (delete-and-extract-region utop-prompt-max (point-max)) utop-history-next)
      ;; Go to after the prompt to insert the previous input
      (goto-char utop-prompt-max)
      ;; Pop one element from history before the cursor and insert it
      (insert (pop utop-history-prev)))))

(defun utop-history-goto-next ()
  "Go to the next entry of the history."
  (interactive)
  (unless (null utop-history-next)
    (with-current-buffer utop-buffer-name
      ;; Push current input before the history cursor
      (push (delete-and-extract-region utop-prompt-max (point-max)) utop-history-prev)
      ;; Go to after the prompt to insert the next input
      (goto-char utop-prompt-max)
      ;; Pop one element from history after the cursor and insert it
      (insert (pop utop-history-next)))))

;; +-----------------------------------------------------------------+
;; | Receiving input from the utop sub-process                     |
;; +-----------------------------------------------------------------+

(defun utop-insert-output (output &optional face)
  "Insert the given output before the prompt."
  (with-current-buffer utop-buffer-name
    (save-excursion
      (let ((line (concat output "\n")))
        ;; Make the line read-only
        (add-text-properties 0 (length line) '(read-only t) line)
        ;; Apply the given face if provided
        (when face (add-text-properties 0 (length line) (list 'face face) line))
        ;; Goto before the prompt
        (goto-char utop-prompt-min)
        ;; Insert the output
        (let ((inhibit-read-only t)) (insert line))
        ;; Advance the prompt
        (setq utop-prompt-min (+ utop-prompt-min (length line)))
        (setq utop-prompt-max (+ utop-prompt-max (length line)))))))

(defun utop-insert-prompt (prompt)
  "Insert the given prompt."
  (with-current-buffer utop-buffer-name
    ;; Make the old prompt sticky so we cannot edit after it
    (let ((inhibit-read-only t))
      (remove-text-properties utop-prompt-min utop-prompt-max '(rear-nonsticky nil)))
    ;; Make the prompt read-only. Make the read-only property
    ;; non-sticky so the buffer can be edited after the prompt
    (utop-add-text-properties-rear-nonsticky 0 (length prompt) '(read-only t) '(face read-only) prompt)
    ;; Goto the end of the buffer
    (goto-char (point-max))
    ;; Make it the start of the prompt
    (setq utop-prompt-min (point))
    ;; Insert the prompt
    (let ((inhibit-read-only t)) (insert prompt))
    ;; Set the end of prompt
    (setq utop-prompt-max (point))))

(defun utop-process-line (line)
  "Process one line from the utop sub-process."
  ;; Extract the command and its argument
  (string-match "\\`\\([a-z-]*\\):\\(.*\\)\\'" line)
  (let ((command (match-string 1 line)) (argument (match-string 2 line)))
    (cond
     ;; Output on stdout
     ((string= command "stdout")
      (utop-insert-output argument 'utop-stdout))
     ;; Output on stderr
     ((string= command "stderr")
      (utop-insert-output argument 'utop-stderr))
     ;; A new prompt
     ((string= command "prompt")
      (let ((prompt (apply utop-prompt ())))
        ;; Check whether there is something to push to the history
        (if (stringp utop-pending)
            ;; Push pending input to the history if it is different
            ;; from the top of the history
            (unless (and (consp utop-history) (string= utop-pending (car utop-history)))
              (push utop-pending utop-history)))
        ;; Clear pending input
        (setq utop-pending nil)
        ;; Reset history
        (setq utop-history-prev utop-history)
        (setq utop-history-next nil)
        ;; Save current prompt
        (setq utop-last-prompt prompt)
        ;; Insert the new prompt
        (utop-insert-prompt prompt)
        ;; Increment the command number
        (setq utop-command-number (+ utop-command-number 1))
        ;; Move the point to the end of buffer in all utop windows if
        ;; needed
        (when utop-move-to-end-of-buffer
          (setq utop-move-to-end-of-buffer nil)
          (let ((buffer (current-buffer)))
            (walk-windows
             (lambda (window)
               (when (eq (window-buffer window) buffer)
                 (select-window window)
                 (goto-char (point-max)))))))))
     ;; Continuation of previous input
     ((string= command "continue")
      ;; Reset history
      (setq utop-history-prev utop-history)
      (setq utop-history-next nil)
      ;; Insert the last prompt
      (utop-insert-prompt utop-last-prompt))
     ;; Complete with a word
     ((string= command "completion-word")
      (insert argument))
     ;; Start of completion
     ((string= command "completion-start")
      (setq utop-completion nil))
     ;; A new possible completion
     ((string= command "completion")
      (setq utop-completion (cons argument utop-completion)))
     ;; End of completion
     ((string= command "completion-stop")
      (with-output-to-temp-buffer "*Completions*"
        (display-completion-list (nreverse utop-completion)))
      (setq utop-completion nil)))))

(defun utop-process-output (process output)
  "Process the output of utop"
  (with-current-buffer utop-buffer-name
    ;; Concatenate the output with the output not yet processed
    (setq utop-output (concat utop-output output))
    ;; Split lines. Each line contains exactly one command
    (let ((lines (split-string utop-output "\n")))
      (while (>= (length lines) 2)
        ;; Process the first line
        (utop-process-line (car lines))
        ;; Remove it and continue
        (setq lines (cdr lines)))
      ;; When the list contains only one element, then this is either
      ;; the end of commands, either an unterminated one, so we save
      ;; it for later
      (setq utop-output (car lines)))))

;; +-----------------------------------------------------------------+
;; | Sending data to the utop sub-process                            |
;; +-----------------------------------------------------------------+

(defun utop-send-input ()
  "Send the text typed at current prompt to the utop
sub-process."
  (interactive)
  (with-current-buffer utop-buffer-name
    ;; Push input to pending input
    (let ((input (buffer-substring-no-properties utop-prompt-max (point-max))))
      (if (stringp utop-pending)
          (setq utop-pending (concat utop-pending "\n" input))
        (setq utop-pending input))
      ;; Goto the end of the buffer
      (goto-char (point-max))
      ;; Terminate input by a newline
      (insert "\n")
      (let ((start utop-prompt-max) (stop (point-max)))
        ;; Make the text read-only
        (add-text-properties start stop '(read-only t))
        ;; Make the old prompt sticky so we cannot edit after it
        (let ((inhibit-read-only t))
          (remove-text-properties utop-prompt-min utop-prompt-max '(rear-nonsticky nil)))
        ;; Makes the text sent read only and add it the frozen face.
        (let ((inhibit-read-only t))
          (add-text-properties start stop '(read-only t face utop-frozen)))
        ;; Move the prompt to the end of the buffer
        (setq utop-prompt-min stop)
        (setq utop-prompt-max stop)
        ;; Send all lines to utop
        (let ((lines (split-string input "\n")))
          (process-send-string utop-process "input:\n")
          (while (consp lines)
            ;; Send the line
            (process-send-string utop-process (concat "data:" (car lines) "\n"))
            ;; Remove it and continue
            (setq lines (cdr lines)))
          (process-send-string utop-process "end:\n"))))))

;; +-----------------------------------------------------------------+
;; | Completion                                                      |
;; +-----------------------------------------------------------------+

(defun utop-complete ()
  "Complete current input."
  (interactive)
  ;; Complete only if the cursor is after the prompt
  (if (>= (point) utop-prompt-max)
      ;; Extract the input before the cursor
      (let ((input (buffer-substring-no-properties utop-prompt-max (point))))
        ;; Split it
        (let ((lines (split-string input "\n")))
          ;; Send all lines to utop
          (process-send-string utop-process "complete:\n")
          (while (consp lines)
            ;; Send the line
            (process-send-string utop-process (concat "data:" (car lines) "\n"))
            ;; Remove it and continue
            (setq lines (cdr lines)))
          (process-send-string utop-process "end:\n")))))

;; +-----------------------------------------------------------------+
;; | Tuareg integration                                              |
;; +-----------------------------------------------------------------+

(defun utop-start ()
  "Start utop if not already started."
  ;; Create the utop buffer if it does not exists, otherwise just
  ;; retreive it
  (let ((buf (get-buffer-create utop-buffer-name)))
    ;; Make it appear
    (display-buffer buf)
    ;; Set the utop mode in that buffer if not already done
    (with-current-buffer buf (unless (eq major-mode 'utop-mode) (utop-mode)))))

(defun utop-eval-region (start end)
  "Eval the current region in utop."
  (interactive "r")
  ;; Start utop if needed
  (save-excursion (utop-start))
  ;; From tuareg
  (setq tuareg-interactive-last-phrase-pos-in-source start)
  ;; Select the text of the region
  (let ((text
         (save-excursion
           ;; Search the start and end of the current paragraph
           (goto-char start)
           (tuareg-skip-blank-and-comments)
           (setq start (point))
           (goto-char end)
           (tuareg-skip-to-end-of-phrase)
           (setq end (point))
           (buffer-substring-no-properties start end))))
    (with-current-buffer utop-buffer-name
      ;; Insert it at the end of the utop buffer
      (goto-char (point-max))
      (insert text ";;")
      ;; Move the point to the end of buffer in all utop windows
      (let ((buffer (current-buffer)))
        (walk-windows
         (lambda (window)
           (when (eq (window-buffer window) buffer)
             (select-window window)
             (goto-char (point-max))))))
      ;; Make sure the cursor is after the prompt when the prompt
      ;; reappear
      (setq utop-move-to-end-of-buffer t)
      ;; Send input to utop now
      (utop-send-input))))

(defun utop-eval-phrase ()
  "Eval the surrounding Caml phrase (or block) in utop."
  (interactive)
  (let ((end))
    (save-excursion
      (let ((pair (tuareg-discover-phrase)))
	(setq end (nth 2 pair))
	(utop-eval-region (nth 0 pair) (nth 1 pair))))
    (if tuareg-skip-after-eval-phrase
	(goto-char end))))

(defun utop-eval-buffer ()
  "Send the buffer to utop."
  (interactive)
  (utop-eval-region (point-min) (point-max)))

;; +-----------------------------------------------------------------+
;; | Edition functions                                               |
;; +-----------------------------------------------------------------+

(defun utop-bol ()
  "Go to the beginning of line or to the end of the prompt."
  (interactive)
  (with-current-buffer utop-buffer-name
    (if (= (point-at-bol) utop-prompt-min)
        (goto-char utop-prompt-max)
      (move-beginning-of-line 1))))

;; +-----------------------------------------------------------------+
;; | The mode                                                        |
;; +-----------------------------------------------------------------+

(defun utop-mode ()
  "Caml Emacs-Lisp Toplevel.

\\{utop-mode-map}"

  ;; Local variables
  (make-local-variable 'utop-mode-map)
  (make-local-variable 'utop-process)
  (make-local-variable 'utop-prompt-min)
  (make-local-variable 'utop-prompt-max)
  (make-local-variable 'utop-last-prompt)
  (make-local-variable 'utop-output)
  (make-local-variable 'utop-command-number)
  (make-local-variable 'utop-history)
  (make-local-variable 'utop-history-prev)
  (make-local-variable 'utop-history-next)
  (make-local-variable 'utop-pending)

  ;; Set the major mode
  (setq major-mode 'utop-mode)
  (setq mode-name "utop")

  ;; Create and use the local keymap utop-mode-map
  (setq utop-mode-map (make-sparse-keymap))
  (use-local-map utop-mode-map)

  ;; Create the sub-process
  (setq utop-process (start-process "utop" (current-buffer) utop-command))

  ;; Filter the output of the sub-process with our filter function
  (set-process-filter utop-process 'utop-process-output)

  ;; Define keys
  (define-key utop-mode-map [return] 'utop-send-input)
  (define-key utop-mode-map [(control ?m)] 'utop-send-input)
  (define-key utop-mode-map [(control ?j)] 'utop-send-input)
  (define-key utop-mode-map [home] 'utop-bol)
  (define-key utop-mode-map [(control ?a)] 'utop-bol)
  (define-key utop-mode-map [(meta ?p)] 'utop-history-goto-prev)
  (define-key utop-mode-map [(meta ?n)] 'utop-history-goto-next)
  (define-key utop-mode-map [tab] 'utop-complete)

  ;; Register the exit hook
  (add-hook 'kill-buffer-hook (lambda () (run-hooks 'utop-exit-hook)) t t)

  ;; Call hooks
  (run-mode-hooks 'utop-mode-hook))

;; +-----------------------------------------------------------------+
;; | Starting utop                                                   |
;; +-----------------------------------------------------------------+

;;;###autoload
(defun utop ()
  "Start utop."
  (interactive)
  ;; Create the utop buffer if it does not exists, otherwise just
  ;; retreive it
  (let ((buf (get-buffer-create utop-buffer-name)))
    ;; Jump to that buffer
    (pop-to-buffer buf)
    ;; Set the utop mode in that buffer if not already done
    (unless (eq major-mode 'utop-mode) (utop-mode))
    ;; Finally return it
    buf))

(provide 'utop)
